---
title: "Challenge 8"
subtitle: "Out of an abundance of caution . . . "
author: Kyeri Kim
output:
  html_document:
    df_print: paged
---
# Abundance Plot Challenge

```{r global_options, include=FALSE}
knitr::opts_chunk$set(echo=TRUE, include=TRUE)
```

## Background
The goal of these challenges are to give you experience with absolute abundance plots, as well as with several features of phyloseq. For several technical reasons, the absolute abundance of amplicon sequencing reads is not biologically meaningful.  However, review of absolute abundance is important for quality control.

## Resources
These challenges are based on the [phyloseq plot_bar tutorial](https://joey711.github.io/phyloseq/plot_bar-examples.html), which will give you starting points for most of the challenges.  

## Data
For all of these challenges you will be using a lemur data phyloseq RDS that is available at `/data/tutorial_data/lemur_subset_phyloseq.rds`.

## Completing and Submitting
You must fill in the chunks below as indicated, but you are free to add other chunks too.  To submit this assignment for full credit you should *commit* and *push*:

1. This file (`challenge8_assignment.Rmd`)
2. The knited version of this file (`challenge8_assignment.html`)

These are the *ONLY* files you should include in your repo.  I strongly recommend that you make a `scratch` subdirectory in your home directory, then make subdirectories for each project in scratch, for example `~/scratch/challenge8`.  If you don't follow this advice, but instead put temporary files in your repo directory, you *must* be sure not to commit them to your repo, you will lose points if you do.



# Challenge Chunks
## Load Libraries
```{r}
library(readr)
library(phyloseq)
library(ggplot2)
library(dplyr)
library(stringr)
library(tidyr)
library(magrittr)
```

## Define Paths
```{r}
poly.rds ="/sharedspace/polystyrene/demultiplexed/dada2/ibiemProject1_subset.rds"
```

## Check Phyloseq RDS
Load your phyloseq RDS and `print` it just to be sure things look sane
```{r}
poly.ps = read_rds(poly.rds)
print(poly.ps)
```

```{r}
# save sample data as a data frame
df <- sample_data(poly.ps)

week <- NULL
collectionID <- NULL
site <- NULL
group <- NULL

# create week and collectionID variables
for(i in 1:nrow(sample_data(poly.ps))){
  week[i] <- strsplit(sample_data(poly.ps)$Description, "_")[[i]][1]
  collectionID[i] <- as.numeric(strsplit(sample_data(poly.ps)$Description, "_")[[i]][2])
}

# create site variable
site[collectionID %in% 1:4] <- "LF"
site[collectionID %in% 5:8] <- "ER"
site[collectionID %in% 9:12] <- "AS"
site[collectionID %in% 13:16] <- "RC"

# create group variable 
group[mod(collectionID, 4) == 1] <- "PS"
group[mod(collectionID, 4) == 2] <- "S"
group[mod(collectionID, 4) == 3] <- "PSS"
group[mod(collectionID, 4) == 0] <- "C"

# add new variables to dataset
df$Week <- as.numeric(week)
df$Site <- site
df$Group <- group

# here is the new sample_data
df
df %>%
  select(Description, Week, Site, Group)

```

Add newly created sample_data file to phyloseq object
```{r}
poly.new.ps <- phyloseq(otu_table(poly.ps),
               df,
               tax_table(poly.ps))

```

Subset by LF
```{r}
ps_LF <- subset_samples(poly.new.ps, Site == "LF")
ps_ER <- subset_samples(poly.new.ps, Site == "ER")
ps_AS <- subset_samples(poly.new.ps, Site == "AS")
ps_RC <- subset_samples(poly.new.ps, Site == "RC")

```

```{r}
# landfill relative abundance plot
rel_abund_func <- function(ps, level){
  top20 <- names(sort(taxa_sums(ps), decreasing=TRUE))[1:20]
ps.top20 <- transform_sample_counts(ps, function(OTU) OTU/sum(OTU))
ps.top20<- prune_taxa(top20, ps.top20)

plot_bar(ps.top20, x="Week", fill= level) + 
  geom_bar(stat = "identity", position = "stack", size=0) +
  facet_wrap(~Group, scales="free_x") 
}
```

```{r}
rel_abund_func(ps_LF, "Family")
rel_abund_func(ps_ER, "Family")
rel_abund_func(ps_AS, "Family")
rel_abund_func(ps_RC, "Family")
```

Plot richness using Shannon
```{r}
plot_richness(poly.new.ps, x="Week", measures=c("Shannon"), color="Group") + 
  theme_bw() +
  facet_wrap(~Site)
```

Plot richness using Simpson
```{r}
# simpson
plot_richness(poly.new.ps, x="Week", measures=c("Simpson"), color="Group") + 
  theme_bw() +
  facet_wrap(~Site)
```

## Relative Abundance Barplot

```{r}
# top 20 taxa across all sites and conditions
top20 <- names(sort(taxa_sums(poly.new.ps), decreasing=TRUE))[1:20]
ps.top20_rel <- transform_sample_counts(poly.new.ps, function(OTU) OTU/sum(OTU))
ps.top20 <- prune_taxa(top20, ps.top20_rel)

# check sample sums all add to 1
sample_sums(ps.top20_rel)

# plot ALL relative abundance in one plot
plot_bar(ps.top20, fill = "Family") +
  labs(y = "Relative Abundance")

# facet by site and group
plot_bar(ps.top20, x="Description", fill="Family") + 
  geom_bar(stat = "identity", position = "stack", size=0) +
  facet_wrap(~Site + Group, scales="free_x") 

```


Plot relative abundance by genus level
```{r}
rel_abund_func_level(ps_LF, "Genus")
rel_abund_func_level(ps_ER, "Genus")
rel_abund_func_level(ps_AS, "Genus")
rel_abund_func_level(ps_RC, "Genus")
```



```{r}
# how to compare increasingness
plot_bar(ps.top20, x="Description", fill="Genus") +
    geom_point(size = 2) +
  geom_line() +
  facet_wrap(~Site, scales="free_x") 
```

```{r}
mdf <-  psmelt(ps.top20)
p <- ggplot(mdf, aes_string(x = "Week", y = "Abundance", color = "Genus")) +
  geom_point(size = 2) +
  geom_line() +
  facet_wrap(~Site, scales="free_x")
p
```

## Permanova

We will prune rare taxa and transform the data
```{r}
sample_min_count = 10000

ps_LF %>%
  prune_samples(sample_sums(.)>=sample_min_count, .) ->
  ps.sample_prune_LF

sample_sums(ps.sample_prune_LF) %>% sort

```


## Basic Challenges
### Simple absolute abundance plot
Start by making a simple absolute abundance bar plot with samples on the x-axis and absolute abundance on the y-axis.  Here is an example from the phyloseq tutorial.
![basic abundance plot](https://joey711.github.io/phyloseq/plot_bar-examples_files/figure-html/unnamed-chunk-3-1.png)
```{r}
plot_bar(poly.ps)
```
```{r}
sn <- sample_names(atacama.ps)
```
```{r}
sample_data(atacama.ps)
```
```{r}
atacama.ps
```



### Absolute abundance plot with kingdom
Now let's look add the *Kingdom* distribution of the reads.  This is a good sanity check for contamination - we expect the *vast* majority of our reads to be bacterial.  There is an example from the phyloseq tutorial doing this at the *genus* level.
![abundance plot with phylogeny](https://joey711.github.io/phyloseq/plot_bar-examples_files/figure-html/unnamed-chunk-4-1.png)
```{r}
plot_bar(atacama.ps, fill="Kingdom")
```

### Absolute abundance plot with phyllum
Now do the same as above, but  coloring by *phyllum* instead of by *kingdom*.
```{r}
plot_bar(atacama.ps, fill="Phylum")
```

## Intermediate Challenges
### Facet Plots
From our plots so far, we see some variation in number of reads per sample.  This is to be expected, but we want to be sure that there are no confounding issues, for example, categories of samples that are outliers.  To this end, we can make facet plots to group samples by different metadata features.  Try to remake the kingdom absolute abundance plot so that samples are grouped by ChowType, this will give us an idea if there is a correlation between chow type and abundance. This plot from the phyloseq tutorial should give you an idea of what you are aiming for:
![abundance plot with phylogeny](https://joey711.github.io/phyloseq/plot_bar-examples_files/figure-html/unnamed-chunk-6-1.png)
```{r}
sample_variables(atacama.ps)
```

```{r}
plot_bar(atacama.ps, facet_grid=~ChowType, fill="Kingdom")
```


Depending on how you do this, you might notice some weird things about your plot:
1. The bars are all black, despite the fact that we trying to color by Kingdom
2. There are spaces between samples

The next challenge will address #1 and we will address #2 in the *Advanced Challeges* section

### Getting rid of annoying black lines
You may have noticed that the bars are divided by short horizontal black lines.  This is because each bar actually consists of a stack of sub-bars, each representing a different OTU.  I generally find the sub-bar delineation to be more confusing than helpful, so I prefer to get rid of it.  By default `plot_bar` outlines each sub-bar in black, but if we outline sub-bars with the same color as the bar fill, we get rid of the delineation.  Try re-making the *Absolute abundance plot with kingdom* from above, but without the black sub-bar outlines.  The section [Add ggplot2 layer to remove the OTU separation lines](https://joey711.github.io/phyloseq/plot_bar-examples.html#add-ggplot2-layer-to-remove-the-otu-separation-lines) should help you do this

```{r}
plot_bar(atacama.ps, fill="Kingdom") +
  geom_bar(aes(color=Kingdom, fill=Kingdom), stat="identity", position="stack")
```

### Facet Plot without annoying black lines
Now try to make the "ChowType" facet plot without the bar outlines.
```{r}
plot_bar(atacama.ps, facet_grid=~ChowType, fill="Kingdom") +
  geom_bar(aes(color=Kingdom, fill=Kingdom), stat="identity", position="stack")
```

## Advanced Challenges
### Missing Values in Facet Plots
By default, the facets of a facet plot use the same scale.  Often this is the right thing to do - if the y-axis of each of the abundance plots had a different scale, it would be hard to compare bars between the facets.  But sometimes it is not, such as our case.  For the ChowType facet plot, since the x-axis is samples, each sample has a place in each facet, even though ChowType is mutually exclusive - each sample corresponds to a lemur, and each lemur only gets one ChowType.  Try to fix the ChowType facet plot so that each facet only shows the samples that correspond to that ChowType.


```{r}
plot_bar(atacama.ps, facet_grid=~ChowType, fill="Kingdom") +
  geom_bar(aes(color=Kingdom, fill=Kingdom), stat="identity", position="stack") +
  facet_grid(~ChowType, scale="free_x")
```

Once you fix that, you will notice that we now have another problem: the facets default to being the same size, despite the number of bars in the facet.  Try adjusting the facet plot so the bars are all the same width, by making the size of the facet proportional to the number bars in it. Hint: check out the `space` argument to `facet_grid`

```{r}
plot_bar(atacama.ps, facet_grid=~ChowType, fill="Kingdom") +
  geom_bar(aes(color=Kingdom, fill=Kingdom), stat="identity", position="stack") +
  facet_grid(~ChowType, scale="free_x", space = "free_x")

```

### Bonus: Abundance Plots without plot_bar
`plot_bar` is pretty convenient, but sometimes you might want more control over plots.  One way is to add ggplot2 layers, as we did above, another is to bypass `plot_bar` and use ggplot2 directly.  Try making one of the above bar plots without `plot_bar`

> Hint: you can see the code for an R function by calling the function without parentheses or arguments

```{r check_plot_bar}
# examining the plot_bar function shows you how it calls ggplot2
plot_bar

```
```{r}
mdf = psmelt(atacama.ps)
```
```{r}
ggplot(mdf, aes_string(x = "Sample", y = "Abundance")) +
  geom_bar(aes(color=Kingdom, fill=Kingdom), stat="identity", position="stack")
```


### Bonus: Get rid of Grid Background
Make one of your plots from above less ugly by getting rid of the annoying grid background
```{r}
ggplot(mdf, aes_string(x = "Sample", y = "Abundance")) +
  geom_bar(aes(color=Kingdom, fill=Kingdom), stat="identity", position="stack")+theme_classic()
```


# Session Info
Always print `sessionInfo` for reproducibility!
```{r}
sessionInfo()
```

